source/extensions/filters/http/istio_stats/istio_stats.cc

```diff
@@ -0,0 +1,670 @@
+    if (it != request_names_.end()) {
+      return it->second;
+    }
+    Stats::StatName name = pool_.add(symbol);
+    request_names_.emplace(symbol, name);
+    return name;
+  }
+
+  Reporter reporter() const { return reporter_; }
+
+  ContextSharedPtr context_;
+  Stats::Scope& scope_;
+  Reporter reporter_;
+  // Backing storage for request strings (lock-free).
+  Stats::StatNameDynamicPool pool_;
+  absl::flat_hash_map<std::string, Stats::StatName> request_names_;
+
+  const bool disable_host_header_fallback_;
+  const std::chrono::milliseconds report_duration_;
+};
+
+using ConfigSharedPtr = std::shared_ptr<Config>;
+
+class IstioStatsFilter : public Http::PassThroughFilter,
+                         public Network::ReadFilter,
+                         public Network::ConnectionCallbacks {
+ public:
+  IstioStatsFilter(ConfigSharedPtr config)
+      : config_(config), context_(*config->context_) {
+    tags_.reserve(25);
+    switch (config_->reporter()) {
+      case Reporter::ServerSidecar:
+        tags_.push_back({context_.reporter_, context_.destination_});
+        break;
+      case Reporter::ClientSidecar:
+        tags_.push_back({context_.reporter_, context_.source_});
+        break;
+    }
+  }
+  ~IstioStatsFilter() { ASSERT(report_timer_ == nullptr); }
+
+  // Http::StreamFilter
+  void onStreamComplete() override {
+    const auto& info = decoder_callbacks_->streamInfo();
+    populatePeerInfo(info, info.filterState());
+    const auto* headers = info.getRequestHeaders();
+    const bool is_grpc =
+        headers && Grpc::Common::isGrpcRequestHeaders(*headers);
+    if (is_grpc) {
+      tags_.push_back({context_.request_protocol_, context_.grpc_});
+    } else {
+      tags_.push_back({context_.request_protocol_, context_.http_});
+    }
+
+    // TODO: copy Http::CodeStatsImpl version for status codes and flags.
+    tags_.push_back(
+        {context_.response_code_,
+         config_->resolve(absl::StrCat(info.responseCode().value_or(0)))});
+    if (is_grpc) {
+      auto response_headers = decoder_callbacks_->responseHeaders();
+      auto response_trailers = decoder_callbacks_->responseTrailers();
+      auto const& optional_status = Grpc::Common::getGrpcStatus(
+          response_trailers
+              ? response_trailers.ref()
+              : *Http::StaticEmptyHeaders::get().response_trailers,
+          response_headers ? response_headers.ref()
+                           : *Http::StaticEmptyHeaders::get().response_headers,
+          info);
+      tags_.push_back({context_.grpc_response_status_,
+                       optional_status ? config_->resolve(absl::StrCat(
+                                             optional_status.value()))
+                                       : context_.empty_});
+    } else {
+      tags_.push_back({context_.grpc_response_status_, context_.empty_});
+    }
+    populateFlagsAndConnectionSecurity(info);
+
+    Stats::Utility::counterFromElements(
```

Comment:
Why don't you use `counterFromStatNames` if all elements are of type _StatName_? It's mentioned to be faster.

Reply:
Thanks, yes I should, changed.